cassandra/concurrent.py

```diff
@@ -74,32 +84,121 @@ def execute_concurrent(session, statements_and_parameters, concurrency=100, rais
-
-    event.wait()
-    if first_error:
-        exc = first_error[0]
+    executor = ConcurrentExecutorGenResults(session, statements_and_parameters) if results_generator else ConcurrentExecutorListResults(session, statements_and_parameters)
+    return executor.execute(concurrency, raise_on_first_error)
+
+
+class _ConcurrentExecutor(object):
+
+    def __init__(self, session, statements_and_params):
+        self.session = session
+        self._enum_statements = enumerate(iter(statements_and_params))
+        self._condition = Condition()
+        self._fail_fast = False
+        self._results_queue = []
+        self._current = 0
+        self._exec_count = 0
+
+    def execute(self, concurrency, fail_fast):
+        self._fail_fast = fail_fast
+        self._results_queue = []
+        self._current = 0
+        self._exec_count = 0
+        with self._condition:
+            for n in xrange(concurrency):
+                if not self._execute_next():
+                    break
+        return self._results()
+
+    def _execute_next(self):
+        # lock must be held
+        try:
+            (idx, (statement, params)) = next(self._enum_statements)
+            self._exec_count += 1
+            self._execute(idx, statement, params)
+            return True
+        except StopIteration:
+            pass
+
+    def _execute(self, idx, statement, params):
+        try:
+            future = self.session.execute_async(statement, params, timeout=None)
+            args = (future, idx)
+            future.add_callbacks(
+                callback=self._on_success, callback_args=args,
+                errback=self._on_error, errback_args=args)
+        except Exception as exc:
+            # exc_info with fail_fast to preserve stack trace info when raising on the client thread
+            # (matches previous behavior -- not sure why we wouldn't want stack trace in the other case)
+            e = sys.exc_info() if self._fail_fast and six.PY2 else exc
+            self._put_result(e, idx, False)
+
+    def _on_success(self, result, future, idx):
+        if future.has_more_pages:
+            result = PagedResult(future, result)
+            future.clear_callbacks()
+        self._put_result(result, idx, True)
+
+    def _on_error(self, result, future, idx):
+        self._put_result(result, idx, False)
+
+    @staticmethod
+    def _raise(exc):
         if six.PY2 and isinstance(exc, tuple):
             (exc_type, value, traceback) = exc
             six.reraise(exc_type, value, traceback)
         else:
             raise exc
-    else:
-        return results
+
+
+class ConcurrentExecutorGenResults(_ConcurrentExecutor):
+
+    def _put_result(self, result, idx, success):
+        with self._condition:
+            heappush(self._results_queue, (idx, (success, result)))
+            self._execute_next()
+            self._condition.notify()
```

Comment:
Does this need to notify every time or could it notify in batches (or when an error happens with fast fail)? Signaling a condition variable is not cheap because it results in multiple `futex` calls (on Linux) and results in several context switches. Because of that expense makes it's advantageous to amortize that cost over multiple request results.  Still thinking of a clean way to implement this...